File: 0647-palindromic-substrings/0647-palindromic-substrings.java
class Solution {
    public int countSubstrings(String s) {
        int len= s.length();
        int result = 0;
        int i = 0;
        while (i < len) {
            int start = i++;
            while (i < len && s.charAt(start) == s.charAt(i)) {
                i++;
            }
			
            int centerLen = i - start;
            result += centerLen * (centerLen + 1) / 2;

            result += expandPalindrome(s, start - 1, i);
        }

        return result;
    }

    private int expandPalindrome(String s, int left, int right) {
        int count = 0;
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
            count++;
        }
        return count;
    }
}

File: 0647-palindromic-substrings/0647-palindromic-substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        
        res = 0
        
        def help(left, right):
            count=0
            while 0 <= left < len(s) and 0 <= right < len(s) and s[left] == s[right]:                
                count += 1
                left -= 1
                right += 1        
            return count
        
        i,n=0, len(s)
        while i<n:
            start=i
            i+=1
            while i<n and s[start]==s[i]:
                i+=1
            repeats =i-start
            res+= repeats *(repeats+1) /2
            res+= help(start-1,i)
            
        return int(res)